File: combinatorics/resource/ST/project_5/bai_toan_7/Postorder.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

// Dinh nghia mot nut cay
struct TreeNode {
    string label;                       // Nhan cua nut
    vector<TreeNode*> children;        // Danh sach con
    TreeNode(string l) : label(l) {}   // Ham tao
};

// Ham duyet hau tu
void postorderTraversal(TreeNode* node) {
    if (!node) return;
    for (TreeNode* child : node->children)
        postorderTraversal(child);     // B1: De quy duyet cac nut con
    cout << node->label << endl;       // B2: Xuat nhan cua nut hien tai
}

// Ham nhap cay tu ban phim
TreeNode* inputTree() {
    string label;
    cout << "Nhap nhan nut: ";
    cin >> label;
    TreeNode* node = new TreeNode(label);
    int n;
    cout << "So con cua " << label << ": ";
    cin >> n;
    for (int i = 0; i < n; ++i) {
        cout << "-- Con thu " << i+1 << " cua " << label << ":\n";
        node->children.push_back(inputTree());
    }
    return node;
}

// Ham main
int main() {
    cout << "Nhap cay:\n";
    TreeNode* root = inputTree();
    cout << "Duyet hau tu:\n";
    postorderTraversal(root);
    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_7/Postorder.py
# Lop dinh nghia cay
class TreeNode:
    def __init__(self, label):
        self.label = label                 # Luu nhan cua nut
        self.children = []                # Danh sach cac nut con

# Ham duyet hau tu
def postorder_traversal(node):
    if node is None:
        return
    for child in node.children:
        postorder_traversal(child)        # B1: Duyet cac nut con truoc
    print(node.label)                     # B2: Xuat nhan cua nut hien tai

# Ham nhap cay tu ban phim
def input_tree():
    label = input("Nhap nhan nut: ")
    node = TreeNode(label)
    n = int(input(f"So con cua {label}: "))
    for i in range(n):
        print(f"-- Con thu {i+1} cua {label}:")
        child = input_tree()
        node.children.append(child)
    return node

# Ham chinh
if __name__ == "__main__":
    print("Nhap cay:")
    root = input_tree()
    print("Duyet hau tu:")
    postorder_traversal(root)